selftests/unit/test_utils_asset.py

```diff
@@ -102,6 +102,32 @@ def test_unknown_scheme(self):
                               None, None, None, [self.cache_dir], None)
         self.assertRaises(asset.UnsupportedProtocolError, invalid.fetch)
 
+    def test_fetch_two_files_same_name_no_hash(self):
+        """
+        Checks that when two different assets which happen to have
+        the same *filename*, are properly stored in the cache
+        directory and that the right one will be given to the user,
+        no matter if a hash is used or not.
+        """
+        second_assetname = self.assetname
+        second_asset_origin_dir = tempfile.mkdtemp(dir=self.basedir)
+        second_asset_local_path = os.path.join(second_asset_origin_dir,
+                                               second_assetname)
+        second_asset_content = 'This is not your first asset content!'
+        with open(second_asset_local_path, 'w') as f:
+            f.write(second_asset_content)
+        second_asset_origin_url = 'file://%s' % second_asset_local_path
+
+        a1 = asset.Asset(self.url, self.assethash, 'sha1', None,
+                         [self.cache_dir], None)
+        a1.fetch()
+        a2 = asset.Asset(second_asset_origin_url, None, None,
+                         [second_asset_origin_dir], [self.cache_dir],
+                         None)
+        a2_path = a2.fetch()
+        with open(a2_path, 'r') as a2_file:
+            self.assertEqual(a2_file.read(), second_asset_content)
```

Comment:
I see 2 main issues of this:

1. it'll just keep growing, once you fetch a file from certain location it'll stay there even when unused (the expire only works when you try to fetch it again) and there are not many ways (except of `find -atime`) to easily examine hundreds of hash dirs.
2. in the end this only decreases the probability of using different asset with the same base-name. For example with clashing url hashes, or simply fetching a dir that clashes by name and contains other files in it. Yes, we can use 64, 128, 256, 2048... characters, but that's just avoiding it.

For those reasons I think it's time to revisit the idea of having a fetched assets database with details. It doesn't have to be a database, it might be just a convention with a way to store the "url" and "hash". That way (unless someone modifies the local disk) we should always get what we asked for.

(by convention I mean something like `$asset` and `$asset.avocado` where in `$asset.avocado` url and hash would be stored plus we could use the `$url_hash/$asset.avocado` to store anonymous downloads)

Reply:
(still could be faked, but definitely not by coincidence)